dkg/pedersen/handler.go

```diff
@@ -513,5 +656,321 @@ func (h *Handler) handleDeal(ctx context.Context, msg types.Deal, out mino.Sende
 		}
 	}
 
+	dela.Logger.Debug().Msgf("%s is certified", h.me)
+
 	return nil
 }
+
+func (h *Handler) announceDkgPublicKey(isCommonNode bool, out mino.Sender, from mino.Address) error {
+	// 6. Send back the public DKG key
+
+	var publicKey kyber.Point
+
+	isOldNode := h.startRes.GetDistKey() != nil
+	isNewNode := !isOldNode
+
+	// if the node is new or a common node it should update its state
+	if isNewNode || isCommonNode {
+		distrKey, err := h.dkg.DistKeyShare()
+		if err != nil {
+			return xerrors.Errorf("failed to get distr key: %v", err)
+		}
+		publicKey = distrKey.Public()
+		// 7. Update the state before sending to acknowledgement to the
+		// orchestrator, so that it can process decrypt requests right away.
+		h.startRes.SetDistKey(distrKey.Public())
+		h.Lock()
+		h.privShare = distrKey.PriShare()
+		h.Unlock()
+	} else {
+		publicKey = h.startRes.GetDistKey()
+	}
+
+	// all the old, new and common nodes should announce their public key to the
+	// initiator, in this way the initiator can make sure that every body has
+	// finished the resharing successfully
+	done := types.NewStartDone(publicKey)
+	err := <-out.Send(done, from)
+	if err != nil {
+		return xerrors.Errorf("got an error while sending pub key: %v", err)
+	}
+
+	dela.Logger.Info().Msgf("%s announced the DKG public key", h.me)
+
+	return nil
+}
+
+// reshare handles the resharing request. Acts differently for the new
+// and old and common nodes
+func (h *Handler) reshare(ctx context.Context, out mino.Sender,
+	from mino.Address, msg types.ResharingRequest, reshares cryChan[types.Reshare], resps cryChan[types.Response]) error {
+
+	addrsNew := msg.GetAddrsNew()
+
+	if len(addrsNew) != len(msg.GetPubkeysNew()) {
+		return xerrors.Errorf("there should be as many participants as pubKey: %d != %d",
+			len(addrsNew), len(msg.GetPubkeysNew()))
+	}
+
+	go func() {
+		err := h.doReshare(ctx, msg, from, out, reshares, resps)
+		if err != nil {
+			h.log.Err(err).Msg("failed to reshare")
+		}
+	}()
+
+	return nil
+}
+
+// doReshare is called when the node has received its reshare message. Note that
+// we might have already received some deals from other nodes in the meantime.
+// The function handles the DKG resharing protocol.
+func (h *Handler) doReshare(ctx context.Context, resharingRequest types.ResharingRequest,
+	from mino.Address, out mino.Sender, reshares cryChan[types.Reshare], resps cryChan[types.Response]) error {
+
+	h.log.Info().Msgf("resharing with %v", resharingRequest.GetAddrsNew())
+
+	addrsNew := resharingRequest.GetAddrsNew()
+
+	isOldNode := h.startRes.GetDistKey() != nil
+	isNewNode := !isOldNode
+
+	// By default the node is not common. Later we check
+	isCommonNode := false
```

Comment:
in this and the other functions that use these variables, wouldn't it make sense to replace this with an enum ? e.g. `nodeDkgGroup` = `old`, `new`, `common` and documenting those constants ?